### PR TITLE
🐛 Improve handling of topology orphaned objects 

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR fixes https://github.com/kubernetes-sigs/cluster-api/issues/10275 and improves how the topology controller handles referenced objects in case of errors, and more specifically
- if InfrastructureCluster is created, but ControlPlane creation fails, InfrastructureCluster is tracked (the issue)
- if infrastructureMachineTemplate is created, but ControlPlane creation fails, infrastructureMachineTemplate is cleaned up
- if infrastructureMachineTemplate is created, but an error happens before MD is created, infrastructureMachineTemplate is cleaned up
- if bootstrapTemplate is created, but an error happens before MD is created, bootstrapTemplate is cleaned up
- if infrastructureMP is created, but an error happens before MP is created, infrastructureMP is cleaned up
- if bootstrapConfig is created, but an error happens before MP is created, bootstrapConfig is cleaned up

I will keep the PR in WIP while I run some additional test

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/10275

/area clusterclass

/cc @sbueringer @chrischdi

### CODE REVIEW

**Comment by mnaser** [comment]
@fabriziopandini : would it be handy to perhaps use `controllerutil.OperationResult` instead of bools (which can maybe be helpful for other things down the line?)

**Comment by fabriziopandini** [comment]
> would it be handy to perhaps use controllerutil.OperationResult instead of bools (which can maybe be helpful for other things down the line?)

controllerutil.OperationResult is not an exact match for reconcileReferencedTemplate, because one possible outcome is a template rotation.

also, this is an internal API, we can eventually refactor it again if we need more things down the line.

**Comment by fabriziopandini** [comment]
Tested reproducing the error with CAPD, no duplicated InfrastructureClusters are created with the fix.
Also tested that after the cluster class patch is fixed, cluster provisioning restarts as expected

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 1]
```diff
@@ -566,28 +598,56 @@ func (r *Reconciler) createMachineDeployment(ctx context.Context, s *scope.Scope
...
+	})
+	if err != nil {
 		return errors.Wrapf(err, "failed to create %s", md.Object.Kind)
 	}
 
+	if createdInfra {
+		infrastructureMachineCleanupFunc = func() {
+			// Best effort cleanup of the InfrastructureMachineTemplate;
+			// If this fails, the object will be garbage collected when the cluster is deleted.
+			_ = r.Client.Delete(ctx, md.InfrastructureMachineTemplate)
```
Comment (chrischdi) [inline internal/controllers/topology/cluster/reconcile_state.go:614]: Maybe worth to the error as a warning? (same for other new deletion funcs)

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 2]
```diff
@@ -1022,44 +1110,44 @@ type reconcileReferencedObjectInput struct {
 // reconcileReferencedObject reconciles the desired state of the referenced object.
 // NOTE: After a referenced object is created it is assumed that the reference should
 // never change (only the content of the object can eventually change). Thus, we are checking for strict compatibility.
-func (r *Reconciler) reconcileReferencedObject(ctx context.Context, in reconcileReferencedObjectInput) error {
+func (r *Reconciler) reconcileReferencedObject(ctx context.Context, in reconcileReferencedObjectInput) (bool, error) {
```
Comment (chrischdi) [inline internal/controllers/topology/cluster/reconcile_state.go:1113]: Maybe worth adding to the godoc that we return true if the object got created?

File: internal/controllers/topology/cluster/reconcile_state_test.go [hunk 3]
```diff
@@ -3221,6 +3412,159 @@ func TestReconcileMachineDeploymentMachineHealthCheck(t *testing.T) {
 	}
 }
 
+func TestReconcileState(t *testing.T) {
+	t.Run("Cluster get reconciled with infrastructure Ref only when reconcileInfrastructureCluster pass and reconcileControlPlane fails ", func(t *testing.T) {
+		g := NewWithT(t)
+
+		currentdCluster := builder.Cluster(metav1.NamespaceDefault, "cluster1").Build()
```
Comment (chrischdi) [inline internal/controllers/topology/cluster/reconcile_state_test.go:3419]:
```suggestion
		currentCluster := builder.Cluster(metav1.NamespaceDefault, "cluster1").Build()
```

**Review by chrischdi — COMMENTED** [review]
/lgtm

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 4]
```diff
@@ -301,61 +320,80 @@ func (r *Reconciler) reconcileInfrastructureCluster(ctx context.Context, s *scop
...
+
+		if createdInfrastructureTemplate {
+			infrastructureMachineCleanupFunc = func() {
+				// Best effort cleanup of the InfrastructureMachineTemplate;
+				// If this fails, the object will be garbage collected when the cluster is deleted.
+				if err := r.Client.Delete(ctx, s.Desired.ControlPlane.InfrastructureMachineTemplate); err != nil {
+					log := tlog.LoggerFrom(ctx).
+						WithValues(s.Desired.ControlPlane.InfrastructureMachineTemplate.GetObjectKind().GroupVersionKind().Kind, s.Desired.ControlPlane.InfrastructureMachineTemplate.GetName()).
+						WithValues("err", err.Error())
+					log.Infof("WARNING! Failed to cleanup InfrastructureMachineTemplate for control plane while handling creation error. The object will be garbage collected when the cluster is deleted.")
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/reconcile_state.go:371]:
```suggestion
					log.Infof("WARNING! Failed to cleanup InfrastructureMachineTemplate for control plane while handling creation or update error. The object will be garbage collected when the cluster is deleted.")
```
Can happen in CP creates & updates

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 5]
```diff
@@ -640,33 +716,68 @@ func (r *Reconciler) updateMachineDeployment(ctx context.Context, s *scope.Scope
...
 
+	if createdBootstrap {
+		bootstrapCleanupFunc = func() {
+			// Best effort cleanup of the BootstrapTemplate;
+			// If this fails, the object will be garbage collected when the cluster is deleted.
+			if err := r.Client.Delete(ctx, desiredMD.BootstrapTemplate); err != nil {
+				log := tlog.LoggerFrom(ctx).
+					WithValues(desiredMD.BootstrapTemplate.GetObjectKind().GroupVersionKind().Kind, desiredMD.BootstrapTemplate.GetName()).
+					WithValues("err", err.Error())
+				log.Infof("WARNING! Failed to cleanup BootstrapTemplate for MachineDeployment while handling upgrade error. The object will be garbage collected when the cluster is deleted.")
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/reconcile_state.go:769]: We sometimes use `handling upgrade` and sometimes `handling update` let's pick one and use it consistently (I would use update)

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 6]
```diff
@@ -676,6 +787,9 @@ func (r *Reconciler) updateMachineDeployment(ctx context.Context, s *scope.Scope
 
 	log.Infof("Patching %s", tlog.KObj{Obj: currentMD.Object})
 	if err := patchHelper.Patch(ctx); err != nil {
+		// Best effort cleanup of the InfrastructureMachineTemplate & BootstrapTemplate (only on creation).
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/reconcile_state.go:790]:
```suggestion
		// Best effort cleanup of the InfrastructureMachineTemplate & BootstrapTemplate (only on template rotation).
```
(same as l.778 I think)

File: internal/controllers/topology/cluster/reconcile_state_test.go [hunk 7]
```diff
@@ -2989,12 +3243,14 @@ func TestReconcileReferencedObjectSequences(t *testing.T) {
...
-					// Execute a reconcile.0
-					g.Expect(r.reconcileReferencedObject(ctx, reconcileReferencedObjectInput{
+					// Execute a reconcile
+					created, err := r.reconcileReferencedObject(ctx, reconcileReferencedObjectInput{
 						cluster: s.Current.Cluster,
 						current: s.Current.ControlPlane.Object,
 						desired: s.Desired.ControlPlane.Object,
-					})).To(Succeed())
+					})
+					g.Expect(err).To(Succeed())
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/reconcile_state_test.go:3252]:
```suggestion
					g.Expect(err).To(NotHaveOccured())
```

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

**Comment by sbueringer** [comment]
Let's cherry-pick (at least into 1.6)

/cherry-pick release-1.6

**Comment by sbueringer** [comment]
@fabriziopandini We probably should cherry-pick manually

**Pull request merged**